llvm/lib/Target/AMDGPU/AMDGPUAliasAnalysis.cpp

```diff
@@ -80,10 +80,13 @@ AliasResult AMDGPUAAResult::alias(const MemoryLocation &LocA,
     } else if (const Argument *Arg = dyn_cast<Argument>(ObjA)) {
       const Function *F = Arg->getParent();
       switch (F->getCallingConv()) {
-      case CallingConv::AMDGPU_KERNEL:
+      case CallingConv::AMDGPU_KERNEL: {
         // In the kernel function, kernel arguments won't alias to (local)
         // variables in shared or private address space.
-        return AliasResult::NoAlias;
+        const auto *ObjB =
+            getUnderlyingObject(B.Ptr->stripPointerCastsForAliasAnalysis());
+        return (ObjA == ObjB) ? AliasResult::MustAlias : AliasResult::NoAlias;
```

Comment:
You can't return MustAlias for an underlying object comparison, as that strips offsets.

Generally, I don't really understand this change or what the actual rule here is supposed to be. The old code makes sense if your IR is UB. If it isn't, then the logic here probably needs to be more along the lines of using the address space of an *identified* underlying object instead of the address space of the original pointer. Basing logic on a comparison of non-identified objects is fishy.

Reply:
I didn't know that about MustAlias, thank you.

My understanding of the intent of the code is that a kernel argument can never alias with a "shared" or "private" address-space-qualified pointer that isn't based on a kernel argument (that's what `(local)` hints at in the comment). Perhaps this is always true for private pointers, as they can never derive from a kernel argument. But proving that a shared pointer is local to a function is going to be tricky; that's why I was trying to use underlying objects.

Another thing I don't understand is whether two distinct shared kernel parameters may alias. In that case there's even more work to be done.

Reply:
> Another thing I don't understand is whether two distinct shared kernel parameters may alias.

Pretty sure they can't. OpenCL requires you to pass a null value to these arguments, so the address is compiler / runtime managed and wouldn't produce an aliasing object

Reply:
Thanks, I've gone with your suggestion.

I have a question, though; is it possible that `ObjA` hits its max lookup during `getUnderlyingObject`, but does in fact alias with `ObjB`? For example, what if it's `ObjB` but with 7+ GEPs, for example? I don't know if this is something to worry about.

Reply:
In this patch we know that ObjA is an Argument, so that shouldn't be possible.

Reply:
Oh, of course.. thanks :+1: